### PR TITLE
API Error Handling and Contexts

### PR DESCRIPTION
**Pull request by manno**
This refers to #347 

It adds contexts to all network calls and makes API error handling less noisy.

### CODE REVIEW

File: acceptance/utilities_test.go [hunk 1]
```diff
@@ -123,9 +123,11 @@ func makeAppWithDir(appName string, instances int, deployFromCurrentDir bool, ap
 
 	// And check presence
 
-	out, err := Epinio("app list", "")
```
Comment (jimmykarily) [inline acceptance/utilities_test.go:126]: oops :D !

File: internal/cli/clients/install_client.go [hunk 2]
```diff
@@ -61,6 +61,8 @@ func (c *InstallClient) Install(cmd *cobra.Command) error {
 	defer log.Info("return")
 	details := log.V(1) // NOTE: Increment of level, not absolute.
 
+	ctx := context.Background()
```
Comment (jimmykarily) [inline internal/cli/clients/install_client.go:64]: What is the purpose of passing a context everywhere, if that context has no way to be cancelled? As far as I understand a context exists as a way to signal various part of the code that what they are doing is no longer relevant (the went "out of context") and as such, they should stop doing it. This is the case with code that is triggered by http requests, when the request is no longer alive, or with code that it triggered from command line when the user does `ctrl+c`. Generally code that should run only as long as something else is still happening.

In this case, the only reason to cancel the context would be if the user stopped the installation (ctrl+c?) or if one of the deployments failed. We would need to create a "child" context with a CancelFunc in order to achieve this (as we do here for example: https://github.com/epinio/epinio/blob/main/internal/api/v1/applications.go#L236) and I don't see this here. Is this meant to happen in another PR?

File: internal/cli/delete.go [hunk 3]
```diff
@@ -14,12 +14,12 @@ var CmdDeleteApp = &cobra.Command{
 	Short: "Deletes an application",
 	Args:  cobra.ExactArgs(1),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		client, err := clients.NewEpinioClient(cmd.Flags())
+		client, err := clients.NewEpinioClient(cmd.Context(), cmd.Flags())
```
Comment (jimmykarily) [inline internal/cli/delete.go:17]: oh cmd has a Context already. I guess this handles the cancelling from the user's side. Maybe we need to handle the case when a deployment fails?

File: acceptance/utilities_test.go (hunk 1 above)
Comment (manno) [reply acceptance/utilities_test.go:126]: I had to move this into an Eventually block, it failed in Github actions with '1/2'. Worked locally, though.

File: deployments/google_services.go [hunk 4]
```diff
@@ -186,15 +188,15 @@ func (k GoogleServices) Deploy(c *kubernetes.Cluster, ui *termui.UI, options kub
...
 		return err
 	}
 
 	return nil
 }
 
-func (k GoogleServices) Upgrade(c *kubernetes.Cluster, ui *termui.UI, options kubernetes.InstallationOptions) error {
+func (k GoogleServices) Upgrade(ctx context.Context, c *kubernetes.Cluster, ui *termui.UI, options kubernetes.InstallationOptions) error {
 	_, err := c.Kubectl.CoreV1().Namespaces().Get(
 		context.Background(),
```
Comment (jimmykarily) [inline deployments/google_services.go:201]: use ctx here?

File: deployments/minibroker.go [hunk 5]
```diff
@@ -151,15 +153,15 @@ func (k Minibroker) Deploy(c *kubernetes.Cluster, ui *termui.UI, options kuberne
...
 		return err
 	}
 
 	return nil
 }
 
-func (k Minibroker) Upgrade(c *kubernetes.Cluster, ui *termui.UI, options kubernetes.InstallationOptions) error {
+func (k Minibroker) Upgrade(ctx context.Context, c *kubernetes.Cluster, ui *termui.UI, options kubernetes.InstallationOptions) error {
 	_, err := c.Kubectl.CoreV1().Namespaces().Get(
 		context.Background(),
```
Comment (jimmykarily) [inline deployments/minibroker.go:166]: here too

**Review by jimmykarily — DISMISSED** [review]

File: internal/cli/clients/install_client.go (hunk 2 above)
Comment (manno) [reply internal/cli/clients/install_client.go]: It's hard to implement a propagated context just in some part of the code base as the shared functions are changed to accept one.

I'm following convention here, by adding a context to all 'request' related funcs. Cobra also provides a top-level context  to commands, which we can propagate.

> A function that is never request-specific may use context.Background(), but err on the side of passing a Context even if you think you don't need to. The default case is to pass a Context; only use context.Background() directly if you have a good reason why the alternative is a mistake.
https://github.com/golang/go/wiki/CodeReviewComments#contexts

I like that this makes request-related funcs much more visible.

Improving cancel-ability we can do in another PR, as discussed.

File: deployments/google_services.go (hunk 4 above)
Comment (manno) [reply deployments/google_services.go]: I fixed all the remaining places and removed TODO() and Background() contexts.

**Pull request merged**